rpc/src/header.rs

```diff
@@ -1,50 +1,197 @@
+        'a: 'fut,
+        'b: 'fut,
+        Self: Sized + Sync + 'fut,
+    {
+        rpc::HeaderClient::header_get_range_by_height(self, from, to)
+    }
 
     /// LocalHead returns the ExtendedHeader of the chain head.
-    #[method(name = "LocalHead")]
-    async fn header_local_head(&self) -> Result<ExtendedHeader, Error>;
+    fn header_local_head<'a, 'fut>(
+        &'a self,
+    ) -> impl Future<Output = Result<ExtendedHeader, Error>> + Send + 'fut
+    where
+        'a: 'fut,
+        Self: Sized + Sync + 'fut,
+    {
+        rpc::HeaderClient::header_local_head(self)
+    }
 
     /// NetworkHead provides the Syncer's view of the current network head.
-    #[method(name = "NetworkHead")]
-    async fn header_network_head(&self) -> Result<ExtendedHeader, Error>;
+    fn header_network_head<'a, 'fut>(
+        &'a self,
+    ) -> impl Future<Output = Result<ExtendedHeader, Error>> + Send + 'fut
+    where
+        'a: 'fut,
+        Self: Sized + Sync + 'fut,
+    {
+        rpc::HeaderClient::header_network_head(self)
+    }
 
     /// Subscribe to recent ExtendedHeaders from the network.
     ///
     /// # Notes
     ///
-    /// Unsubscribe is not implemented by Celestia nodes.
-    #[subscription(name = "Subscribe", unsubscribe = "Unsubscribe", item = ExtendedHeader)]
-    async fn header_subscribe(&self) -> SubscriptionResult;
+    /// If client returns [`Error::HttpNotImplemented`], the subscription will fallback to
+    /// using [`HeaderClient::header_wait_for_height`] for streaming the headers.
+    fn header_subscribe<'a>(
+        &'a self,
+    ) -> Pin<Box<dyn Stream<Item = Result<ExtendedHeader, Error>> + Send + 'a>>
+    where
+        Self: SubscriptionClientT + Sized + Sync,
+    {
+        try_stream! {
+            let mut head = rpc::HeaderClient::header_local_head(self).await?;
+            head.validate().map_err(custom_client_error)?;
+
+            let subscription_res = rpc::HeaderSubscriptionClient::header_subscribe(self).await;
+            let has_real_sub = !matches!(&subscription_res, Err(Error::HttpNotImplemented));
+
+            let mut subscription = if has_real_sub {
+                Some(subscription_res?)
+            } else {
+                None
+            };
+
+            loop {
+                let header = if has_real_sub {
+                    subscription
+                        .as_mut()
+                        .expect("must be some")
+                        .next()
+                        .await
+                        .ok_or_else(|| custom_client_error("unexpected end of stream"))??
+                } else {
+                    rpc::HeaderClient::header_wait_for_height(self, head.height().value() + 1).await?
+                };
+
+                header.validate().map_err(custom_client_error)?;
+                head.verify_adjacent(&header).map_err(custom_client_error)?;
+
+                head = header.clone();
+                yield header;
+            }
+        }
```

Comment:
```suggestion
        try_stream! {
            let mut head = rpc::HeaderClient::header_local_head(self).await?;
            head.validate().map_err(custom_client_error)?;

            let mut real_subscription = match rpc::HeaderSubscriptionClient::header_subscribe(self).await {
                Ok(subscription) => Ok(Some(subscription)),
                Err(Error::HttpNotImplemented) => Ok(None),
                Err(e) => Err(e)
            }?;

            loop {
                let header = match &mut real_subscription {
                    Some(subscription) => subscription.next().await
                        .ok_or_else(|| custom_client_error("unexpected end of stream"))??,
                    None => rpc::HeaderClient::header_wait_for_height(self, head.height().value() + 1).await?,
                };

                header.validate().map_err(custom_client_error)?;
                head.verify_adjacent(&header).map_err(custom_client_error)?;

                head = header.clone();
                yield header;
            }
        }
```
similar point to above